source/xep/mam/MamManager.cpp

```diff
@@ -107,135 +115,140 @@ void MamManager::requestArchiveForJid(const QString& jid, const QString &last)
 
         Swift::JID sJid(jid.toStdString());
 
-        //qDebug() << xw.getXmlResult();
-
-        client_->getIQRouter()->sendIQ(Swift::IQ::createRequest(Swift::IQ::Set, sJid, msgId,
-                                                                std::make_shared<Swift::RawXMLPayload>(xw.getXmlResult().toStdString())
-                                                                ));
+        // Calling getXmlResult flushes content and subsequent calls are empty! 
+        // qDebug() << xw.getXmlResult(); 
 
+        MamRequest::ref mamRequest = MamRequest::create(Swift::IQ::Set, sJid, xw.getXmlResult().toStdString(), client_->getIQRouter());
+        mamRequest->onResponse.connect(boost::bind(&MamManager::processFinIq, this, _1, _2, _3));
+        mamRequest->send();
     }
 }
 
-// FIXME rewrite me!
-// this fails sometimes :-(.
-// use custom payload parser to filter out mam messages!
-
```

Comment:
This is a nice approach. I also tried this, but did not get it working as it should. Give me some time to try this implementation with some MaM data :-).